src/error.rs

```diff
@@ -162,3 +162,60 @@ impl From<untrusted::EndOfInput> for Error {
         Error::BadDer
     }
 }
+
+// Return the more specific of two errors.
+pub(crate) fn most_specific_error(prev: Error, new: Error) -> Error {
+    // Assign an error a numeric value ranking it by specificity.
+    #[allow(clippy::as_conversions)] // We won't exceed u32 errors.
+    fn rank(e: Error) -> u32 {
+        match e {
+            // Errors related to certificate validity
+            Error::CertNotValidYet | Error::CertExpired => 100027,
+            Error::CertNotValidForName => 100026,
+            Error::CertRevoked => 100025,
+            Error::InvalidSignatureForPublicKey => 100024,
+            Error::SignatureAlgorithmMismatch => 100023,
+            Error::RequiredEkuNotFound => 100022,
+            Error::NameConstraintViolation => 100021,
+            Error::PathLenConstraintViolated => 100020,
+            Error::CaUsedAsEndEntity | Error::EndEntityUsedAsCa => 100019,
+            Error::IssuerNotCrlSigner => 100018,
+
+            // Errors related to supported features used in an invalid way.
+            Error::InvalidCertValidity => 100017,
+            Error::InvalidNetworkMaskConstraint => 100016,
+            Error::InvalidSerialNumber => 100015,
+            Error::InvalidCrlNumber => 100014,
+
+            // Errors related to unsupported features.
+            Error::UnsupportedSignatureAlgorithmForPublicKey => 100013,
+            Error::UnsupportedSignatureAlgorithm => 100012,
+            Error::UnsupportedCriticalExtension => 100011,
+            Error::UnsupportedCrlVersion => 100010,
+            Error::UnsupportedDeltaCrl => 10009,
+            Error::UnsupportedIndirectCrl => 10008,
+            Error::UnsupportedRevocationReason => 10007,
+
+            // Errors related to malformed data.
+            Error::MalformedDnsIdentifier => 10006,
+            Error::MalformedNameConstraint => 10005,
+            Error::MalformedExtensions => 10004,
+            Error::ExtensionValueInvalid => 10003,
+
+            // Generic DER errors.
+            Error::BadDerTime => 10002,
+            Error::BadDer => 100001,
+
+            // Default catch all error - should be renamed in the future.
+            Error::UnknownIssuer => 0,
+
+            // For other errors, use error's position in the enum.
+            e => (e as u32) + 1,
```

Comment:
i think we might have all the errors listed above -- so removing this line would mean we get an error when a new error is added that forces us to consider the rank of the new error. that seems preferable to quietly ranking new errors as being above `UnknownIssuer` but below everything else?

Reply:
Yeah, we can rely on the `match` requiring exhaustiveness, so I don't think we need this?

Reply:
Great idea. This also surfaced that I had missed including `UnsupportedCertVersion`. All fixed up.